### PR TITLE
Run performance tests in CI

### PR DESCRIPTION
**Pull request by tillarnold**
@vakaras  So far this is just the x.py function to run the benchmarks.

### CODE REVIEW

File: x.py [hunk 1]
```diff
@@ -257,6 +261,55 @@ def ide(args):
...
+
+    env["PRUSTI_SERVER_ADDRESS"]="localhost:12345"
+    try:
+        with open('benchmarks.csv') as csv_file:
+            csv_reader = csv.reader(csv_file, delimiter=',')
+            for row in csv_reader:
+                file_path = row[0]
+                results[file_path] = []
+                report("Starting to benchmark {}", row)
+                for i in range(warmup_iterations):
```
Comment (vakaras) [inline x.py]: I would suggest moving the warm-up outside of the `csv_reader` loop.

File: x.py [hunk 2]
```diff
@@ -257,6 +261,55 @@ def ide(args):
...
 
+def run_benchmarks(args):
+    warmup_iterations = 3
+    bench_iterations = 10
+    results = {}
+
+    report("running benchmarks")
+    env = get_env()
+    report("Starting prusti-server")
+    server_process = subprocess.Popen(["./target/release/prusti-server-driver","--port","12345"], env=env)
```
Comment (vakaras) [inline x.py]: The `prusti-server-driver` path is probably platform-specific, so would be good to add a way to support multiple OS versions.

**Review by vakaras — CHANGES_REQUESTED** [review]
Looks mostly good. Could you please also add a corresponding GitHub action?

**Comment by vakaras** [comment]
@aterga @fpoli @ArquintL Why does http://viper.ethz.ch/downloads/ViperToolsNightlyMac.zip have both `/boogie/Binaries/boogie` and `/boogie/Binaries/Boogie`, which conflict on a case-insensitive file system?

File: x.py (hunk 2 above)
Comment (fpoli) [reply x.py]: Also, the binary to be executed should be `prusti-server` and it should not need to set environment variables.

File: x.py [hunk 3]
```diff
@@ -257,6 +261,55 @@ def ide(args):
...
+    json_result = json.dumps(results, indent = 2)
+    timestamp = time.time()
+    with open("benchmark-output/benchmark" + str(timestamp) + ".json", "w") as outfile: 
+        outfile.write(json_result) 
+
+                
+
+def measure_time(path, env):
+    start_time = time.perf_counter()
+    run_command(["./target/debug/prusti-rustc","--edition=2018", path],env=env)
```
Comment (fpoli) [inline x.py]: Have you tried running this without setting the `env`? I think it should work.

File: x.py [hunk 4]
```diff
@@ -257,6 +261,55 @@ def ide(args):
...
+        os.makedirs('benchmark-output')
+
+    json_result = json.dumps(results, indent = 2)
+    timestamp = time.time()
+    with open("benchmark-output/benchmark" + str(timestamp) + ".json", "w") as outfile: 
+        outfile.write(json_result) 
+
+                
+
+def measure_time(path, env):
```
Comment (fpoli) [inline x.py]: This function name looks a a bit too generic, what about `measure_prusti_time(input_path, env)`?

File: x.py (hunk 3 above)
Comment (fpoli) [reply x.py]: I mean, passing an env that just contains `PRUSTI_SERVER_ADDRESS`, and not the `get_env()`.

File: x.py (hunk 2 above)
Comment (vakaras) [reply x.py]: @fpoli How do you kill it then properly? The kill below kills `prusti-server`, but `prusti-server-driver` keeps running.

**Comment by tillarnold** [comment]
@vakaras regarding the GitHub action: which branch should I push the results to? There might be problems with pushing to `gh-pages` because the action of `gh-pages-src` would probably overwrite that.

File: x.py (hunk 2 above)
Comment (fpoli) [reply x.py]: Oh, I didn't notice that. `prusti-server` should propagate SIGINT to `prusti-server-driver` or do something to shutdown properly. I opened issue #249. It's also possible that `z3` instances keeps running, even when using `prusti-server-driver`.

**Comment by fpoli** [comment]
> @aterga @fpoli @ArquintL Why does http://viper.ethz.ch/downloads/ViperToolsNightlyMac.zip have both `/boogie/Binaries/boogie` and `/boogie/Binaries/Boogie`, which conflict on a case-insensitive file system?

I don't know. It may be a bug of the Jenkins build that prepares the zip.

**Comment by vakaras** [comment]
> @vakaras regarding the GitHub action: which branch should I push the results to? There might be problems with pushing to `gh-pages` because the action of `gh-pages-src` would probably overwrite that.

I think it would be best to push to `gh-pages` and change `gh-pages-src` to overwrite only specific directories. @Aurel300 Do you have any insights?

**Comment by Aurel300** [comment]
> I think it would be best to push to `gh-pages` and change `gh-pages-src` to overwrite only specific directories. @Aurel300 Do you have any insights?

Sure, I guess we can simply enable [`keep-files`](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-keeping-existing-files-keep_files) on the `gh-pages-src` action. Pushing to `-src` from benchmarks is messy IMO.

**Comment by ArquintL** [comment]
> @aterga @fpoli @ArquintL Why does http://viper.ethz.ch/downloads/ViperToolsNightlyMac.zip have both `/boogie/Binaries/boogie` and `/boogie/Binaries/Boogie`, which conflict on a case-insensitive file system?

I only see `boogie`. Is there a `Boogie` file other than `Boogie.exe`?

**Comment by fpoli** [comment]
> I only see `boogie`. Is there a `Boogie` file other than `Boogie.exe`?

Yes

File: .github/workflows/test.yml [hunk 5]
```diff
@@ -68,6 +68,36 @@ jobs:
         cd prusti-contracts-test/
         cargo build
 
+  benchmark:
+    strategy:
+      matrix:
+        os: [ macos-latest, ubuntu-latest, windows-latest ]
```
Comment (vakaras) [inline .github/workflows/test.yml]: I think it would be fine to run benchmarks on `ubuntu-latest` only.

**Comment by vakaras** [comment]
> > I think it would be best to push to `gh-pages` and change `gh-pages-src` to overwrite only specific directories. @Aurel300 Do you have any insights?
> 
> Sure, I guess we can simply enable [`keep-files`](https://github.com/peaceiris/actions-gh-pages#%EF%B8%8F-keeping-existing-files-keep_files) on the `gh-pages-src` action. Pushing to `-src` from benchmarks is messy IMO.

Do I understand it correctly that with `keep-files=false`, the action overrides only the `destination_dir`? So, one option would be to:
1. change `gh-pages-src` to write not to root, but to a subdirectory;
2. add https://viperproject.github.io/ with a proper index page that links to Viper stuff and to a Prusti index page in a subdirectory.

Any concerns regarding this design?

**Comment by Aurel300** [comment]
> Do I understand it correctly that with `keep-files=false`, the action overrides only the `destination_dir`?

Yes.

With `keep-files: true`, I think it merges the directory structure, overriding any files with the same filename but otherwise leaving everything intact.

> Any concerns regarding this design?

A little bit, yes. `https://viperproject.github.io/` would need to be a separate repository, but I guess that's fine. We could set the action in `gh-pages-src` to write to e.g. `prusti-dev/dev-guide` and `prusti-dev/user-guide`, but then it is not writing the index, so that would need to be done manually in `gh-pages`. A bit of a pain and I don't see the point.

So let's just change the GH action in `gh-pages-src` to also have `keep-files: true`.

**Comment by tillarnold** [comment]
@vakaras the CI push action fails with `remote: Permission to viperproject/prusti-dev.git denied to github-actions[bot].` is this only because this is running in the context of this PR and thus does not have push access to the repository or is there something I need to change?

**Comment by vakaras** [comment]
> @vakaras the CI push action fails with `remote: Permission to viperproject/prusti-dev.git denied to github-actions[bot].` is this only because this is running in the context of this PR and thus does not have push access to the repository or is there something I need to change?

One problem is that this job is configured to run on each pull request and each should not:

```yaml
on:
  push:
    branches: 'master'
  pull_request:
    branches: [ '**' ]
```

Could you move it to `benchmarks.yaml` and configure to run only on push to `master`?

**Review by vakaras — APPROVED** [review]
LGTM. Thanks for implementing this!

**Pull request merged**